Домашнее задание 3. Телефонная книга

DaniarFF opened the pull request.

DaniarFF committed:
Домашнее задание 3. Телефонная книга

yaporyvaev commented on PhoneBook/AbonentList.cs:
```diff
@@ -0,0 +1,23 @@
+    public class AbonentList
+    {
+        [JsonPropertyName("abonents")]
+        public List<Abonent> Abonent { get; set; }
```
Назвать Abonents. У тебя же список

Dezodemius replied:
реально

yaporyvaev commented on PhoneBook/AbonentList.cs:
```diff
@@ -0,0 +1,23 @@
+{
+    public class AbonentList
+    {
+        [JsonPropertyName("abonents")]
```
Настроить через конвеншены, убрать атрибуты

yaporyvaev commented on PhoneBook/AbonentList.cs:
```diff
@@ -0,0 +1,23 @@
+    }
+    public class Abonent
+    {
+        [JsonPropertyName("name")]
```
Убрать атрибут

yaporyvaev commented on PhoneBook/PhoneBookClient.cs:
```diff
@@ -0,0 +1,101 @@
+using System.Text;
+using System.Text.Json;
+using System.Threading.Tasks;
+using static System.Runtime.InteropServices.JavaScript.JSType;
```
Для чего это? Надо убрать

yaporyvaev commented on PhoneBook/PhoneBookClient.cs:
```diff
@@ -0,0 +1,101 @@
+using System.Text.Json;
+using System.Threading.Tasks;
+using static System.Runtime.InteropServices.JavaScript.JSType;
+using System.Xml.Linq;
```
Откуда xml?

yaporyvaev commented on PhoneBook/PhoneBookClient.cs:
```diff
@@ -0,0 +1,101 @@
+{
+    public class PhoneBookClient
+    {
+        PhoneBookRepository phoneBookRepository = new PhoneBookRepository();
```
Явно указать модификатор доступа. Приватные поля класса начинаются с "_", например "_phoneBookRepository".
Сделать инициализацию через конструктор

yaporyvaev commented on PhoneBook/PhoneBookClient.cs:
```diff
@@ -0,0 +1,101 @@
+
+namespace PhoneBook
+{
+    public class PhoneBookClient
```
Почему класс называется Client? Client'ами называются классы, которые обращаются во внешние системы. В твоем случае это PhoneBookService

yaporyvaev commented on PhoneBook/PhoneBookClient.cs:
```diff
@@ -0,0 +1,101 @@
+    {
+        PhoneBookRepository phoneBookRepository = new PhoneBookRepository();
+
+        public async Task GetAllAbonents(string path)
```
Сервис не должен принимать путь до файла. Сервис ничего не должен знать о доступе к данным, это ответственность DAL - слоя доступа к данным. В твоем сценарии, это репозиторий, хоть он и не выделен в отдельный слой

yaporyvaev commented on PhoneBook/PhoneBookClient.cs:
```diff
@@ -0,0 +1,101 @@
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine("Что то пошло не так:( ");
```
Что-то*

yaporyvaev replied:
Убрать консоль

yaporyvaev commented on PhoneBook/PhoneBookClient.cs:
```diff
@@ -0,0 +1,101 @@
+            }
+        }
+
+        public async Task CreateNewAbonent(string path, AbonentList abonentList)
```
Убрать path

yaporyvaev replied:
Метод называется "Создать нового абонента", но принимает список абонентов, сделать название более релевантным, либо исправить параметры

yaporyvaev replied:
Зачем здесь abonentList?

yaporyvaev replied:
Данные об абоненте должны приходить в параметрах

yaporyvaev commented on PhoneBook/PhoneBookClient.cs:
```diff
@@ -0,0 +1,101 @@
+                AbonentList list = await phoneBookRepository.GetAllAbonents(path);
+                foreach (var item in list.Abonent)
+                {
+                    Console.WriteLine($"Имя: {item.Name} Телефон: {item.PhoneNumber}");
```
Класс Console разрешается использовать только в Program.cs. Представь, что твой сервис и репозиторий, находится не в консольном приложении, а в библиотеке. В библиотеке нет консоли, соответственно ты не сможешь её вызвать. По хорошему, у тебя классы PhoneBookClient и PhoneBookRepository должны быть в разных проектах

yaporyvaev commented on PhoneBook/PhoneBookClient.cs:
```diff
@@ -0,0 +1,101 @@
+        {
+            try 
+            {
+                Console.WriteLine("Напиши имя: ");
```
Убрать консоль

yaporyvaev commented on PhoneBook/PhoneBookClient.cs:
```diff
@@ -0,0 +1,101 @@
+                Console.WriteLine("Напиши телефон: ");
+                int.TryParse(Console.ReadLine(), out int number);
+                {
+                    if(number < 0 || number > 1000000000000) { Console.WriteLine("Неверное кол - во цифр"); } //лень было обрабатывать формат, надеюсь не нужно было
```
Если номер будет начинаться со знака "+", упадёт. Подумать как еще сделать валидацию

yaporyvaev commented on PhoneBook/PhoneBookClient.cs:
```diff
@@ -0,0 +1,101 @@
+            }        
+        }
+
+        public async Task DeleteAbonent( string path, AbonentList abonentList)
```
Убрать параметры. Принимать данные для удаления

yaporyvaev commented on PhoneBook/PhoneBookClient.cs:
```diff
@@ -0,0 +1,101 @@
+        {
+            try
+            {
+                Console.WriteLine("Напиши имя: ");
```
Убрать консоль

yaporyvaev commented on PhoneBook/PhoneBookClient.cs:
```diff
@@ -0,0 +1,101 @@
+                Console.WriteLine("Напиши имя: ");
+                string name = Console.ReadLine();
+
+                await phoneBookRepository.DeleteAbonent(name, path, abonentList);
```
Передавать только идентификатор, по которому удалять

yaporyvaev commented on PhoneBook/PhoneBookClient.cs:
```diff
@@ -0,0 +1,101 @@
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine("Что то пошло не так:( ");
```
Консоль

yaporyvaev commented on PhoneBook/PhoneBookClient.cs:
```diff
@@ -0,0 +1,101 @@
+            }
+        }
+
+        public async Task FindAbonent(string path, AbonentList abonentList)
```
Убрать параметры, принимать идентификатор для поиска

yaporyvaev commented on PhoneBook/PhoneBookClient.cs:
```diff
@@ -0,0 +1,101 @@
+
+        public async Task FindAbonent(string path, AbonentList abonentList)
+        {
+            Console.WriteLine("Напиши имя или телефон");
```
Консоль

yaporyvaev commented on PhoneBook/PhoneBookClient.cs:
```diff
@@ -0,0 +1,101 @@
+            Console.WriteLine("Напиши имя или телефон");
+            string name = Console.ReadLine();
+
+            if (int.TryParse(name, out int number))  // Возможа ли такая обработка? Или это бред
```
Если хочешь удаление по фио и по телефону -- сделай 2 разных метода

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,67 @@
+{
+    public class PhoneBookRepository
+    {
+        public async Task<AbonentList> GetAllAbonents(string path) 
```
Не принимать path извне

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,67 @@
+    {
+        public async Task<AbonentList> GetAllAbonents(string path) 
+        {
+            using (FileStream fs = new FileStream(path, FileMode.OpenOrCreate))
```
using в новых шарпах можно упросить

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,67 @@
+
+        public async Task CreateNewAbonent(string name, long number, string path, AbonentList abonentList) 
+        {
+            Abonent newAbonent = new Abonent() 
```
Можно принимать сразу Abontent

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,67 @@
+                PhoneNumber = number,
+            };
+
+            if(FindAbonent(name, path, abonentList).Result == null) 
```
.Result?

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,67 @@
+            }
+            else 
+            {
+                Console.WriteLine("Абонент уже существует");
```
Консоль

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,67 @@
+        public async Task DeleteAbonent(string name, string path, AbonentList abonentList)
+        {
+            var itemToDelete = abonentList.Abonent.FirstOrDefault(s => s.Name == name);
+            abonentList.Abonent.Remove(itemToDelete);
```
FirstOrDefault может вернуть null. Нужна проверка перед удалением

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,67 @@
+            using (FileStream fs = new FileStream(path, FileMode.Create))
+            {
+                await JsonSerializer.SerializeAsync<AbonentList>(fs, abonentList);
+                Console.WriteLine($"Абонент {name} Удален");
```
Консоль

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,67 @@
+            }
+        }
+
+        public async Task<Abonent> FindAbonent(string name, string path, AbonentList abonentList)
```
Если регистр аргумента и регистр строки в базе не совпадет, то не найдет получается

yaporyvaev commented on PhoneBook/Program.cs:
```diff
@@ -0,0 +1,45 @@
+    {
+        static async Task Main(string[] args)
+        {
+            string path = "C:\\Users\\Данияр\\source\\repos\\CSharpEducation\\PhoneBook\\PhoneBook.json";
```
Принимать из конфиг файла

DaniarFF committed:
1. Настроил работу с JSON по соглашению с помощью опций

2. Переименовал классы и методы на более подходящие
3. Создал App.setting для хранения локальных данных
4. Исправил параметры методов, убрал параметры пути и списка абонентов
5. Убрал работу с консолью вне класса Program.cs
6. Настроил валидацию номера с помощью регулярных выражений
7. Валидация имени по регистру первой буквы

yaporyvaev commented on .gitignore:
```diff
@@ -396,3 +396,4 @@ FodyWeavers.xsd
 
 # JetBrains Rider
 *.sln.iml
+/PhoneBook/App.config
```
Зачем конфиг в игнор добавил? Как человеку, который склонировал репозиторий запускать проект без этого файла?

yaporyvaev replied:
????????????????????????????

yaporyvaev commented on PhoneBook/Abonents.cs:
```diff
@@ -0,0 +1,14 @@
+    {
+        public List<Abonent> Abonent { get; set; }
+    }
+    public class Abonent
```
Каждый класс в отдельный файл надо вынести

Dezodemius replied:
и документацию добавить

yaporyvaev commented on PhoneBook/Abonents.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace PhoneBook
+{
+    public class Abonents
```
Не понимаю смысл этого класса

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -1,67 +1,96 @@
     public class PhoneBookRepository
     {
-        public async Task<AbonentList> GetAllAbonents(string path) 
+        readonly string _filePath;
```
Модификатор доступа

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -1,67 +1,96 @@
-        public async Task<AbonentList> GetAllAbonents(string path) 
+        readonly string _filePath;
+
+        JsonSerializerOptions options = new JsonSerializerOptions
```
Модификатор доступа. Приватное поле класса начинается с "_"

Dezodemius replied:
по нашему соглашению не начинается

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -1,67 +1,96 @@
-            Abonent newAbonent = new Abonent() 
-            { 
+            using FileStream fs = new FileStream(_filePath, FileMode.OpenOrCreate);
+            Abonents abonents = await JsonSerializer.DeserializeAsync<Abonents>(fs, options);
```
Можно просто десериализовать в List<Abonent> и не нужен будет бесполезный класс Abonents

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -1,67 +1,96 @@
-            if(FindAbonent(name, path, abonentList).Result == null) 
-            {
-                abonentList.Abonent.Add(newAbonent);
+            if (await FindWithName(name) == null)
```
Лучше вынести вызов метода на отдельную строку

yaporyvaev replied:
FindByName, а не With

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -1,67 +1,96 @@
+
+                var itemToDelete = abonents.Abonent.FirstOrDefault(s => s.Name == name);
+                if (itemToDelete != null) 
+                {
```
Зачем выбрасывать ошибку при удалении? Можно сказать что всё ок, т.к. записи не было, ничего не удалилось и никому при этом плохо не стало

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -1,67 +1,96 @@
         }
-        public async Task<Abonent> FindAbonent(long number, string path, AbonentList abonentList)
+
+        public async Task<Abonent> FindWithNumber(string number)
```
FindByPhoneNumber

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -1,67 +1,96 @@
+            return abonents.Abonent.FirstOrDefault(s => s.PhoneNumber == number);
+        }
+
+        public string FirstLetterToUpperCase(string str)
```
Плохое название метода, нет глагола

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -1,67 +1,96 @@
+
+                name = FirstLetterToUpperCase(name);
+
+                var itemToDelete = abonents.Abonent.FirstOrDefault(s => s.Name == name);
```
Это не будет нормально работать. Если ты добавил изначально "ДАнияр", а потом пытаешься удалить "Данияр" у тебя будет ошибка

yaporyvaev commented on PhoneBook/PhoneBookService.cs:
```diff
@@ -0,0 +1,90 @@
+{
+    public class PhoneBookService
+    {
+        static string _filePath = ConfigurationManager.AppSettings["path"];
```
Обращение к файлу конфигурации не должно тут быть, только в корне приложения. Лучше принимать в конструкторе уже сконфигурированный PhoneBookRepository. Это называется Dependency Injection

yaporyvaev commented on PhoneBook/PhoneBookService.cs:
```diff
@@ -0,0 +1,90 @@
+            }
+            catch (Exception ex)
+            {
+                throw new Exception(ex.Message);
```
Написал выше

yaporyvaev commented on PhoneBook/PhoneBookService.cs:
```diff
@@ -0,0 +1,90 @@
+                {
+                    return await _phoneBookRepository.FindWithNumber(searchString);      
+                }
+
```
Лишний пропуск строки

yaporyvaev commented on PhoneBook/PhoneBookService.cs:
```diff
@@ -0,0 +1,90 @@
+            Regex regex = new Regex(@"^\+7\d{10}$");
+            if (regex.IsMatch(searchString)) 
+            {
+                try
```
Достаточно одного блока try\catch на метод

yaporyvaev commented on PhoneBook/PhoneBookService.cs:
```diff
@@ -0,0 +1,90 @@
+            }
+            catch (Exception ex)
+            {
+                throw new Exception(ex.Message);
```
Тебе уже приходит объект типа Exception извне, зачем ты создаешь новый? Если ты хочешь пробросить ошибку дальше, то можно просто написать throw;
И если ты ничего не делаешь в блоке catch, а только пробрасываешь исключение дальше, то try\catch можно вообще не писать

yaporyvaev commented on PhoneBook/Program.cs:
```diff
@@ -25,21 +22,87 @@ static async Task Main(string[] args)
+                Console.WriteLine("Напиши телефон: ");
+                string number = Console.ReadLine();
+
+                if (await phoneBookService.AddNewAbonent(name, number))
```
Вынести вызов метода на отдельную строку

yaporyvaev commented on PhoneBook/Program.cs:
```diff
@@ -25,21 +22,87 @@ static async Task Main(string[] args)
+                Console.WriteLine("Напиши имя: ");
+                string name = Console.ReadLine();
+
+                if (await phoneBookService.DeleteAbonent(name))
```
Написал выше

yaporyvaev commented on PhoneBook/PhoneBookService.cs:
```diff
@@ -0,0 +1,90 @@
+                }
+                else 
+                { 
+                    throw new Exception("Неправильный формат телефона, должен быть в формате +79998887766");
```
В этом кейсе неправильно использовать дефолтный Exception. По-хорошему, надо создать какой-нибудь ValidationException, унаследовать его от Exception и пользоваться им. Но нам пока пофиг

yaporyvaev commented on PhoneBook/Program.cs:
```diff
@@ -25,21 +22,87 @@ static async Task Main(string[] args)
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex.Message);
```
В ex.Message может быть системный текст, непонятный пользователю, стоит его ему показывать?

Dezodemius commented on PhoneBook/PhoneBook.csproj:
```diff
@@ -0,0 +1,16 @@
+    <Nullable>enable</Nullable>
+  </PropertyGroup>
+
+  <ProjectExtensions><VisualStudio><UserProperties /></VisualStudio></ProjectExtensions>
```
шо это?

Dezodemius commented on PhoneBook/PhoneBook.json:
Для тестов можно и коммитить, но такое лучше не стоит в репозиторий добавлять

Dezodemius commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,96 @@
+            WriteIndented = true
+        };
+
+        public PhoneBookRepository(string path)
```
По соглашению нужно конструктор в конце файла ставить

Dezodemius replied:
Кажется, лучше чтоб имя и путь к файлу были внутри зашиты. Чтоб снаружи никто не знал про него так просто

Dezodemius commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,96 @@
+
+        public async Task<Abonents> GetAll()
+        {
+            using FileStream fs = new FileStream(_filePath, FileMode.OpenOrCreate);
```
Я бы в конструкторе это делал. А getAll пусть просто возвращает все элементы из abonents

Dezodemius commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,96 @@
+            return abonents;
+        }
+
+        public async Task AddNew(string name, string number)
```
Можно просто Add

Dezodemius replied:
а чё он async? Зачем?

Dezodemius commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,96 @@
+
+        public async Task AddNew(string name, string number)
+        {
+            Abonents abonents = await GetAll(); 
```
было б проще коллекцию завести в классе и к ней обращаться, чтоб каждый раз в файл не записывать

Dezodemius commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,96 @@
+            }
+            else
+            {
+                throw new Exception("Абонент уже существует");
```
ну будто бы какое-то получше исключение подобрать надо. А то Exception слишком жирно

Dezodemius commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,96 @@
+            return abonents.Abonent.FirstOrDefault(s => s.PhoneNumber == number);
+        }
+
+        public string FirstLetterToUpperCase(string str)
```
а нафиг он нужон вообще? Метод этот. Вот как пользователя назвали, так пусть и добавляет

Dezodemius commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,96 @@
+            await JsonSerializer.SerializeAsync<Abonents>(fs, abonents, options);
+        }
+
+        public async Task Delete(string name)
```
а если у тебя будет 10 Егоров в абонентах, то все удалятся? Я бы привязывался к номеру телефона, а не к имени. Потому что у номер один владелец только может быть

DaniarFF committed:
Изменения:

1. Внедрен Dependency Injection
2. Исправление ошибок
3. Рефракторинг кода на основе соглашения Directum

yaporyvaev commented on PhoneBook/Models/Abonent.cs:
```diff
@@ -0,0 +1,12 @@
+    /// </summary>
+    public class Abonent
+    {
+        public string Name { get; set; }
```
Комментарий

yaporyvaev commented on PhoneBook/PhoneBook.json:
```diff
@@ -0,0 +1,30 @@
+[
```
Не надо этот файл в репозиторий добавлять мне кажется

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,115 @@
+    /// </summary>
+    public class PhoneBookRepository : IPhoneBookRepository
+  {
+    #region Поля и свойства
```
втф

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,115 @@
+
+    #endregion
+
+    #region Вложенные типы
```
сейм

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,115 @@
+
+    #region Вложенные типы
+
+    JsonSerializerOptions options = new JsonSerializerOptions()
```
Модификатор доступа

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,115 @@
+        PhoneNumber = number,
+      };
+
+      Abonent abonent = await FindByName(name);
```
Тёзки лесом идут?

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,115 @@
+      }
+      else
+      {
+        throw new InvalidOperationException("Абонент уже существует");
```
Дефолтные эксепшены не принято выкидывать в любой непонятной ситуации, тут лучше свой эксепшен сделать

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,115 @@
+
+        name = CapitalizeFirstLetter(name);
+
+        var itemToDelete = abonents.FirstOrDefault(s => s.Name == name);
```
Шо если у людей имя одинаковое?

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,115 @@
+      }
+      catch (Exception ex)
+      {
+        throw new Exception("Ошибка удаления");
```
Не нравится. Ты ничего не делаешь в блоке catch полезного. Наоборот, ты перетираешь исходный объект, в котором была вся информация об ошибке, не надо так

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,115 @@
+      }
+    }
+
+    public async Task<Abonent> FindByName(string name)
```
Как будто по имени может найтись несколько контактов

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,115 @@
+      return abonents.FirstOrDefault(s => s.PhoneNumber == number);
+    }
+
+    public string CapitalizeFirstLetter(string str)
```
В чем прекол этого метода?

yaporyvaev commented on PhoneBook/PhoneBookService.cs:
```diff
@@ -0,0 +1,77 @@
+  {
+    #region Поля и свойства
+
+    private string filePath;
```
Не нужно здесь

yaporyvaev commented on PhoneBook/PhoneBookService.cs:
```diff
@@ -0,0 +1,77 @@
+    #endregion
+
+    #region Методы
+    public async Task<List<Abonent>> GetAllAbonents()
```
Асинк авейт не нужен

yaporyvaev commented on PhoneBook/PhoneBookService.cs:
```diff
@@ -0,0 +1,77 @@
+      }
+      else
+      {
+        throw new Exception("Неправильный формат телефона, должен быть в формате +79998887766");
```
ValidationException должен быть

yaporyvaev replied:
Вообще плохо строить логику на эксепшенах, но это нам пока не важно, просто держи в голове

yaporyvaev commented on PhoneBook/PhoneBookService.cs:
```diff
@@ -0,0 +1,77 @@
+      if (regex.IsMatch(number))
+      {
+        await phoneBookRepository.Add(name, number);
+        return true;
```
Зачем тут bool результат?

yaporyvaev commented on PhoneBook/PhoneBookService.cs:
```diff
@@ -0,0 +1,77 @@
+    public async Task<bool> DeleteAbonent(string name)
+    {
+      await phoneBookRepository.Delete(name);
+      return true;
```
Зачем bool результат?

yaporyvaev commented on PhoneBook/Program.cs:
```diff
@@ -0,0 +1,123 @@
+  {
+    static async Task Main(string[] args)
+    {
+      IServiceCollection services = new ServiceCollection();
```
А тут шо регионы свои бесполезные не воткнул

yaporyvaev commented on PhoneBook/Program.cs:
```diff
@@ -0,0 +1,123 @@
+      }
+    }
+
+    static async Task GetAllAbonents(IPhoneBookService phoneBookService)
```
модификатор доступа

yaporyvaev replied:
Аргумент лучше вынести в поле

yaporyvaev commented on PhoneBook/Program.cs:
```diff
@@ -0,0 +1,123 @@
+    {
+      try
+      {
+        List<Abonent> list = await phoneBookService.GetAllAbonents();
```
Если с типами данных, объявлением и инциализацией переменных разобрался, используй var, если понимаешь шо это такое и как работает

yaporyvaev commented on PhoneBook/Program.cs:
```diff
@@ -0,0 +1,123 @@
+      }
+    }
+
+    static async Task AddNew(IPhoneBookService phoneBookService)
```
модификатор доступа

yaporyvaev commented on PhoneBook/Settings/App.config:
```diff
@@ -0,0 +1,6 @@
+﻿<?xml version="1.0" encoding="utf-8" ?>
```
Что за файл, зачем он?

yaporyvaev commented on PhoneBook/Settings/ApplicationSettings.cs:
```diff
@@ -0,0 +1,18 @@
+    /// </summary>
+    public string RepositoryPath { get; set; }
+
+    public ApplicationSettings(string repositoryPath)
```
Конфиг файлам не нужен конструктор

DaniarFF committed:
Изменения:

1. Исправлены ошибки
2. Оптимизирована логика приложения, каждый класс теперь делает то, что должен
3. Убран ненужный метод для изменения строки
4. Созданы кастомные Exceptions

yaporyvaev commented on .gitignore:
```diff
@@ -396,3 +396,4 @@ FodyWeavers.xsd
 
 # JetBrains Rider
 *.sln.iml
+/PhoneBook/App.config
```
Удоли

yaporyvaev commented on PhoneBook/Exceptions/ContactAlreadyExistsException.cs:
```diff
@@ -0,0 +1,18 @@
+﻿namespace PhoneBook.Exceptions
+{
+  public class ContactAlreadyExistsException : ApplicationException
```
Слишком конкретизировал. Достаточно сделать общий ValidationException

yaporyvaev commented on PhoneBook/Exceptions/ContactDoesntExistException.cs:
```diff
@@ -0,0 +1,18 @@
+﻿namespace PhoneBook.Exceptions
+{
+  public class ContactDoesntExistException : ApplicationException
```
Соответственно этот уже не нужен

yaporyvaev commented on PhoneBook/Exceptions/InvalidPhoneNumberFormatException.cs:
```diff
@@ -0,0 +1,20 @@
+{
+  public class InvalidPhoneNumberFormatException : ValidationException
+  {
+    public InvalidPhoneNumberFormatException(string message) : base(message)
```
Тоже не нужен

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,87 @@
+
+    public async Task Add(Contact contact)
+    {
+      try
```
Что даёт тебе блок trycatch в этом случае?

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,87 @@
+
+    public async Task Delete(Contact contact)
+    {
+      try
```
Писал выше

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,87 @@
+      {
+        List<Contact> contacts = GetAll().ToList();
+
+        foreach (var item in contacts)
```
Зачем цикл?

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,87 @@
+
+        foreach (var item in contacts)
+        {
+          if (item.PhoneNumber == contact.PhoneNumber)
```
Если из аргумента используешь только телефон, зачем передавал целый объект?

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,87 @@
+        {
+          if (item.PhoneNumber == contact.PhoneNumber)
+          {
+            contact = item;
```
Переписывать аргумент метода это вообще смерть. Это ссылочный тип, если ты его перепишешь здесь, он у тебя изменится там, откуда ты его передавал и можно вообще в ногу себе выстрелить, если не делаешь это специально

yaporyvaev commented on PhoneBook/PhoneBookRepository.cs:
```diff
@@ -0,0 +1,87 @@
+      }
+    }
+
+    public async Task Delete(Contact contact)
```
Я бы предложил пересмотреть реализацию этого метода

yaporyvaev commented on PhoneBook/PhoneBookService.cs:
```diff
@@ -0,0 +1,96 @@
+
+    public async Task<Contact> AddNewContact(string name, string number)
+    {
+      Regex regex = new Regex(@"^\+7\d{10}$");
```
Лучше опустить ниже по коду, к использованию. Создаешь объект, который может никогда не использоваться

yaporyvaev commented on PhoneBook/PhoneBookService.cs:
```diff
@@ -0,0 +1,96 @@
+      if (regex.IsMatch(number))
+      {
+        await phoneBookRepository.Add(contactToAdd);
+        return default;
```
Шо за дефолт

yaporyvaev commented on PhoneBook/PhoneBookService.cs:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    public async Task DeleteContact(string number)
+    {
```
Форматирование поправить

yaporyvaev commented on PhoneBook/Program.cs:
```diff
@@ -0,0 +1,147 @@
+  {
+    static async Task Main(string[] args)
+    {
+      var cfg = new ConfigurationBuilder().AddJsonFile("config.json").Build();
```
Где сам файл конфига? Его нет в репо

yaporyvaev replied:
Этот файл обычно называется appsettings.json

DaniarFF committed:
Исправлены ошибки
